presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/StateCompiler.java

```diff
@@ -211,93 +206,96 @@ else if (stackType == Slice.class) {
         return null;
     }
 
-    private static <T> void generateDeserialize(ClassDefinition definition, Class<T> clazz, List<StateField> fields)
+    private static <T> void generateDeserialize(ClassDefinition definition, CallSiteBinder binder, Class<T> clazz, List<StateField> fields)
     {
         Parameter block = arg("block", Block.class);
         Parameter index = arg("index", int.class);
         Parameter state = arg("state", Object.class);
         MethodDefinition method = definition.declareMethod(a(PUBLIC), "deserialize", type(void.class), block, index, state);
-
         BytecodeBlock deserializerBody = method.getBody();
-
+        Scope scope = method.getScope();
         if (fields.size() == 1) {
-            Method setter = getSetter(clazz, fields.get(0));
-            Method blockGetter = getBlockGetter(setter.getParameterTypes()[0]);
-            deserializerBody.append(state.cast(setter.getDeclaringClass()).invoke(setter, invokeStatic(blockGetter, block, index)));
+            StateField field = fields.get(0);
+            Method setter = getSetter(clazz, field);
+            if (!field.isPrimitiveType()) {
+                deserializerBody.append(new IfStatement()
+                        .condition(block.invoke("isNull", boolean.class, index))
+                        .ifTrue(state.cast(setter.getDeclaringClass()).invoke(setter, constantNull(field.getType())))
+                        .ifFalse(state.cast(setter.getDeclaringClass()).invoke(setter, constantType(binder, field.getSqlType()).getValue(block, index))));
+            }
+            else {
+                // For primitive type, we need to cast here because byte state has a different serialized type (long).
```

Comment:
I think this is better:

```// For primitive type, we need to cast here because we serialize byte fields with TINYINT (whose java type is long).```

Ditto for the other 3 occurrences.